### PR TITLE
Add --ide-ledger flag to daml script cli

### LINKED ISSUE #16532: Add IdeLedger support to daml script cli tool
Required for #16525, specifically the incremental build tests.
Likely have an option that specifies we're using the IDE ledger, and forbid `ledger-host`, `ledger-port`, `participant-config`

### PR DESCRIPTION
**Pull request by samuel-williams-da**
Allows us to run single or all tests in a dar fully locally, via the IdeLedgerClient.
Some flag combinations have been restricted to allow this.
Resolves #16532

### CODE REVIEW

**Comment by samuel-williams-da** [comment]
Looking deeper at the code here, I think it could do with a fairly large refactor - the whole --all and --script-name seems far too split up, when most of the logic should be the same.
We can consider merging them at some point in the future

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/TestMain.scala [hunk 1]
```diff
@@ -49,36 +75,43 @@ object TestMain extends StrictLogging {
...
+      config.ledgerMode match {
+        case LedgerMode.ParticipantConfig(file) =>
+          val source = Source.fromFile(file)
+          val fileContent =
+            try {
+              source.mkString
+            } finally {
+              source.close
+            }
+          val jsVal = fileContent.parseJson
```
Comment (remyhaemmerle-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/TestMain.scala:93]:
```suggestion
          val jsVal = java.nio.file.Files.readString(file.toPath).parseJson
```

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerMode.scala [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+package com.daml.lf.engine.script
+
+import java.io.File
+
+sealed abstract class LedgerMode extends Product with Serializable
+object LedgerMode {
+  final case class LedgerAddress(host: String, port: Int) extends LedgerMode
+  final case class ParticipantConfig(file: File) extends LedgerMode
+  final case class IdeLedger() extends LedgerMode
+}
```
Comment (remyhaemmerle-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerMode.scala:13]: Nice.

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/RunnerMain.scala [hunk 3]
```diff
@@ -15,15 +15,19 @@ import scala.io.Source
...
 import com.daml.lf.language.Ast.Package
+import com.daml.lf.language.Ast.Type
 import com.daml.lf.typesig.EnvironmentSignature
 import com.daml.lf.typesig.reader.SignatureReader
 import com.daml.grpc.adapter.{AkkaExecutionSequencerPool, ExecutionSequencerFactory}
 import com.daml.auth.TokenHolder
 import com.daml.ledger.api.tls.TlsConfiguration
 import com.daml.lf.data.NoCopy
+import com.daml.lf.engine.script.ledgerinteraction.IdeLedgerClient
 import com.daml.lf.engine.script.ledgerinteraction.ScriptTimeMode
```
Comment (remyhaemmerle-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/RunnerMain.scala:31]:
```suggestion
import com.daml.lf.engine.script.ledgerinteraction.{IdeLedgerClient,ScriptTimeMode}
```

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/TestMain.scala [hunk 4]
```diff
@@ -49,36 +75,43 @@ object TestMain extends StrictLogging {
     implicit val materializer: Materializer = Materializer(system)
     implicit val executionContext: ExecutionContext = system.dispatcher
 
-    val (participantParams, participantCleanup) = config.participantConfig match {
```
Comment (remyhaemmerle-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/TestMain.scala:52]: Do I read properly that `participantCleanup` was never doing anything ?

**Review by remyhaemmerle-da — APPROVED** [review]
Nice work.

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/TestMain.scala (hunk 4 above)
Comment (samuel-williams-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/TestMain.scala:52]: Ah right yeah, I had a comment for it previously, but forgot to commit that - yes, it was always set to no-op

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerMode.scala (hunk 2 above)
Comment (samuel-williams-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerMode.scala:13]: Thoughts on the names btw? I was a little unsure, as I'm not super familiar with what a participant config actually is. Especially given the first LedgerAddress case just builds a Participant of the same type as ParticipantConfig

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/TestMain.scala (hunk 1 above)
Comment (samuel-williams-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/TestMain.scala:93]: Ah nice! Will do

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/TestMain.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/TestMain.scala:52]: :+1:

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerMode.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerMode.scala:13]: Not fantastic: naming is hard. 
I think it is ok if you always use them with prefix (`LedgerMode.XXX`). 
maybe LedgerConfig is better ?

anyway it is better now than before.

Comment (samuel-williams-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerMode.scala:13]: I'll give them another think, to be sure
Maybe `RemoteParticipantHost`, `RemoteParticipantConfig` and `IdeLedgerParticipant`

**Comment by samuel-williams-da** [comment]
Windows CI failing with `'msys2' install failed previously. Please uninstall it and try again.`, problem with the agent. Waiting until fixed

**Comment by samuel-williams-da** [comment]
See https://github.com/digital-asset/docs.daml.com/pull/208 for docs

**Pull request merged**